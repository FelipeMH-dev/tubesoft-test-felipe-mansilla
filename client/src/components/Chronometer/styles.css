.hover:hover {
  background-color: rgb(230, 28, 136) !important;
}

.btn-OFF {
  background-color: gray !important;
  color: black;
  display: inline-block;
  font-weight: 400;
  text-align: center;
  vertical-align: middle;
  cursor: default  !important;
  -webkit-user-select: none;
  border: 1px solid transparent;
  padding: 0.375rem 0.75rem;
  font-size: 0.9375rem;
  line-height: 1.5;
  border-radius: 0.25rem;
}

.card-body {
  height: 430px;
}
