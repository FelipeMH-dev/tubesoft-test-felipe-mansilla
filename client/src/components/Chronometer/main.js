import React, { useEffect, useState, useRef } from "react";
import "./styles.css";
import imgTime from "../../images/time.svg";
import axios from "axios";

export default function Chronometer() {
  const [hours, setHours] = useState(0);
  const [minutes, setMinutes] = useState(0);
  const [seconds, setSeconds] = useState(0);

  let intervalRef = useRef();
  const increaseSeconds = () => setSeconds((prev) => prev + 1);
  const increaseMinutes = () => setMinutes((prev) => prev + 1);
  const increaseHours = () => setHours((prev) => prev + 1);

  const [alert, setAlert] = useState(false);
  const [message, setMessage] = useState("");
  const [finishTimestamp, setFinish] = useState("");
  const [startTimestamp, setStart] = useState(true);

  const btnStart = useRef(null);

  useEffect(() => {
    console.log(seconds);
    if (seconds === 59) {
      setSeconds(0);
      increaseMinutes();
    }
    if (minutes === 59) {
      setMinutes(0);
      increaseHours();
    }
  }, [minutes, seconds]);

  const Start = () => {
    if (startTimestamp) {
      setAlert(false);
      intervalRef.current = setInterval(increaseSeconds, 1000);
      setStart(false);
      btnStart.current.className = "btn btn-OFF";
      return () => clearInterval(intervalRef.current);
    }
  };

  const Pause = () => {
    setAlert(false);
    setStart(true);
    btnStart.current.className = "btn btn-primary hover";
    clearInterval(intervalRef.current);
  };

  const Reset = () => {
    setSeconds(0);
    setMinutes(0);
    setHours(0);
  };
  
  const Finish = async () => {
    let timestamp = `${hours > 9 ? hours.toString() : "0" + hours.toString()}:${
      minutes > 9 ? minutes.toString() : "0" + minutes.toString()
    }:${seconds > 9 ? seconds.toString() : "0" + seconds.toString()}`;

    try {
      const res = await axios.post("http://localhost:4000/api/timestamps", {
        timestamp: timestamp,
      });
      setStart(true);
      btnStart.current.className = "btn btn-primary hover";
      setAlert(true);
      setMessage(res.data.message);
      setFinish(res.data.data.timestamp);
      clearInterval(intervalRef.current);
      Reset();
    } catch (e) {
      console.log(e);
    }
  };

  return (
    <div className="container ">
      <div className="row justify-content-center">
        <div className="col-md-4">
          <div className="card  ">
            <div className="card-header bg-success text-white">
              <div className="row">
                <div className="col-md-8">Chronometer</div>
                <div className="col-md-4 ">
                  <img
                    src={imgTime}
                    className="img-fluid mx-auto"
                    alt="imgQuestion"
                  />
                </div>
              </div>

              <div className="mt-4">
                <div className="container">
                  <div className="bg-white text-dark my-2 p-2 rounded text-center">
                    <h2>
                      {hours > 9 ? "" : 0}
                      {hours} : {minutes > 9 ? "" : 0}
                      {minutes} : {seconds > 9 ? "" : 0}
                      {seconds}
                    </h2>
                  </div>
                </div>

                <div className="mt-4">
                  <div className="container">
                    <div className="row">
                      <div className="col-md-4 text-white">
                        <button
                          ref={btnStart}
                          id="start"
                          className="btn btn-primary hover"
                          onClick={() => Start()}
                        >
                          Start
                        </button>
                      </div>
                      <div className="col-md-4 text-white">
                        <button
                          id="pause"
                          className="btn btn-primary hover"
                          onClick={() => Pause()}
                        >
                          Pause
                        </button>
                      </div>
                      <div className="col-md-4 text-white">
                        <button
                          id="finish"
                          className="btn btn-primary hover"
                          onClick={() => Finish()}
                        >
                          Finish
                        </button>
                      </div>
                    </div>
                  </div>
                </div>

                <div className="container">
                  {alert && (
                    <div className="bg-info text-white my-2 p-2 rounded text-center">
                      {message}
                      <br />
                      {finishTimestamp}
                    </div>
                  )}
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  );
}
