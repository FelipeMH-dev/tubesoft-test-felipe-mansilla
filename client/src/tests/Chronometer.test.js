import React from "react";
import { shallow } from "enzyme";
import Chronometer from "../components/Chronometer/main";

jest.setTimeout(10000);
describe("Test Chronometer", () => {
  let wrapper = shallow(<Chronometer />);

  test("Snapshot <Chronometer/>", () => {
    expect(wrapper).toMatchSnapshot();
  });
  test("test start", async () => {
    wrapper.find("#start").simulate("click");

    await new Promise((res) =>
      setTimeout(() => {
        const counterText = wrapper.find("h2").text().trim();
        console.log("Start: ", counterText);
        expect(counterText).not.toBe("00 : 00 : 00");
        res();
      }, 1000)
    );
  });

  test("test pause", async () => {
    await new Promise((res) =>
      setTimeout(() => {
        wrapper.find("#pause").simulate("click");
        const counterText = wrapper.find("h2").text().trim();
        console.log("Pause in: ", counterText);
        res();
      }, 3000)
    );

    await new Promise((res) =>
      setTimeout(() => {
        const counterText = wrapper.find("h2").text().trim();
        console.log("Check pause after two seconds: ", counterText);
        expect(counterText).toBe("00 : 00 : 04");
        res();
      }, 2000)
    );
  });

  test("test finish", async () => {
    const counterText = wrapper.find("h2").text().trim();
    console.log("Time before finish: ", counterText);
    wrapper.find("#finish").simulate("click");
    await new Promise((res) =>
      setTimeout(() => {
        const counterText = wrapper.find("h2").text().trim();
        console.log("Finish and Reset: ", counterText);
        expect(counterText).toBe("00 : 00 : 00");
        res();
      }, 1000)
    );
  });
});
