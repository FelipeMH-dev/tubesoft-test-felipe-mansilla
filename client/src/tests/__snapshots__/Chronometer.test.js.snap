// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Chronometer Snapshot <Chronometer/> 1`] = `
<div
  className="container "
>
  <div
    className="row justify-content-center"
  >
    <div
      className="col-md-4"
    >
      <div
        className="card  "
      >
        <div
          className="card-header bg-success text-white"
        >
          <div
            className="row"
          >
            <div
              className="col-md-8"
            >
              Chronometer
            </div>
            <div
              className="col-md-4 "
            >
              <img
                alt="imgQuestion"
                className="img-fluid mx-auto"
                src="time.svg"
              />
            </div>
          </div>
          <div
            className="mt-4"
          >
            <div
              className="container"
            >
              <div
                className="bg-white text-dark my-2 p-2 rounded text-center"
              >
                <h2>
                  0
                  0
                   : 
                  0
                  0
                   : 
                  0
                  0
                </h2>
              </div>
            </div>
            <div
              className="mt-4"
            >
              <div
                className="container"
              >
                <div
                  className="row"
                >
                  <div
                    className="col-md-4 text-white"
                  >
                    <button
                      className="btn btn-primary"
                      id="start"
                      onClick={[Function]}
                    >
                      Start
                    </button>
                  </div>
                  <div
                    className="col-md-4 text-white"
                  >
                    <button
                      className="btn btn-primary"
                      id="pause"
                      onClick={[Function]}
                    >
                      Pause
                    </button>
                  </div>
                  <div
                    className="col-md-4 text-white"
                  >
                    <button
                      className="btn btn-primary"
                      id="finish"
                      onClick={[Function]}
                    >
                      Finish
                    </button>
                  </div>
                </div>
              </div>
            </div>
            <div
              className="container"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
